File: python/200_Number_of_Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    self.BFS(grid, [i, j])
                    count += 1
                    
        return count
    
    
    def BFS(self, grid: List[List[str]], startPos: List[int]):
        directions = [ [-1,  0],
                       [0,  1],
                       [1, 0],
                       [0, -1] ]
        
        grid[startPos[0]][startPos[1]] = "0"
        
        queue = deque()
        queue.append(startPos)
        
        while queue:
            curPos = queue.popleft()
            
            for direction in directions:
                curRow = curPos[0] + direction[0]
                curCol = curPos[1] + direction[1]
                
                if curRow < 0 or curRow >= len(grid) or curCol < 0 or curCol >= len(grid[0]) or grid[curRow][curCol] == "0":
                    continue  
                
                queue.append([curRow, curCol])
                grid[curRow][curCol] = "0"

File: swift/200_number_of_islands.swift
struct DirectionIterator: IteratorProtocol {
  private let directions: [(row: Int, col: Int)] = [
    (row: 1, col: 0),
    (row: -1, col: 0),
    (row: 0, col: -1),
    (row: 0, col: 1),
  ]
  private var curIdx: Int = 0

  mutating func next() -> (row: Int, col: Int)? {
    guard curIdx < directions.count else { return nil }
    let direction = directions[curIdx]
    curIdx += 1
    return direction
  }
}

class Solution {
  func numIslands(_ grid: [[Character]]) -> Int {
    var seenLands: [[Bool]] = Array(
      repeating: Array(repeating: false, count: grid[0].count), count: grid.count)
    var islandCnt: Int = 0

    for row in 0..<grid.count {
      for col in 0..<grid[0].count {
        if grid[row][col] == "1" && !seenLands[row][col] {
          print(row, col)
          findLands(from: (row, col), grid: grid, seenLands: &seenLands)
          islandCnt += 1
        }
      }
    }

    return islandCnt
  }

  private func findLands(
    from pos: (row: Int, col: Int), grid: [[Character]], seenLands: inout [[Bool]]
  ) {
    var landQueue: [(row: Int, col: Int)] = []
    landQueue.append(pos)

    while !landQueue.isEmpty {
      let curLandPos = landQueue.removeFirst()
      let curRow: Int = curLandPos.row
      let curCol: Int = curLandPos.col

      var directionIterator = DirectionIterator()

      while let direction = directionIterator.next() {
        let newRow: Int = curRow + direction.row
        let newCol: Int = curCol + direction.col

        guard 0 <= newRow && newRow < grid.count && 0 <= newCol && newCol < grid[0].count else {
          continue
        }

        if grid[newRow][newCol] == "1" && !seenLands[curRow][curCol] {
          landQueue.append((newRow, newCol))
        }
      }

      seenLands[curRow][curCol] = true
    }
  }
}
